projects/RabbitMQ.Client/client/api/ICredentialsRefresher.cs

```diff
@@ -80,65 +84,123 @@ public ICredentialsProvider Register(ICredentialsProvider provider, ICredentials
-            if (_registrations.TryRemove(provider, out System.Timers.Timer? timer))
+            lock (_lockObj)
             {
-                try
+                if (_registrations.TryGetValue(provider, out var registration))
                 {
+                    _registrations.Remove(provider);
+
                     TimerBasedCredentialRefresherEventSource.Log.Unregistered(provider.Name);
-                    timer.Stop();
-                }
-                finally
-                {
-                    timer.Dispose();
+                    registration.Dispose();
+                    return true;
                 }
-                return true;
-            }
-            else
-            {
-                return false;
             }
+
+            return false;
         }
 
-        private System.Timers.Timer scheduleTimer(ICredentialsProvider provider, ICredentialsRefresher.NotifyCredentialRefreshedAsync callback)
+        private class TimerRegistration : IDisposable
         {
-            System.Timers.Timer timer = new System.Timers.Timer();
-            timer.Interval = provider.ValidUntil!.Value.TotalMilliseconds * (1.0 - (1 / 3.0));
-            timer.Elapsed += (o, e) =>
+
+            private readonly object _lockObj;
+            private System.Timers.Timer? _timer;
+            private bool _disposed;
+
+            public ICredentialsRefresher.NotifyCredentialRefreshedAsync Callback { get; set; }
+
+            public TimerRegistration(object lockObj, ICredentialsRefresher.NotifyCredentialRefreshedAsync callback)
             {
-                TimerBasedCredentialRefresherEventSource.Log.TriggeredTimer(provider.Name);
+                _lockObj = lockObj;
+                Callback = callback;
+            }
+
+            public void ScheduleTimer(ICredentialsProvider provider)
+            {
+                if (provider.ValidUntil == null)
+                {
+                    throw new ArgumentNullException("ValidUntil of " + nameof(provider) + " was null");
+                }
+                if (_disposed)
+                {
+                    throw new InvalidOperationException("Registration already disposed");
+                }
+
+                var newTimer = new Timer();
+                newTimer.Interval = provider.ValidUntil.Value.TotalMilliseconds * (1.0 - 1 / 3.0);
+                newTimer.Elapsed += (o, e) =>
+                {
+                    TimerBasedCredentialRefresherEventSource.Log.TriggeredTimer(provider.Name);
+
+                    lock (_lockObj)
+                    {
+                        try
+                        {
+                            if (_disposed)
+                            {
+                                // We were waiting and the registration has been disposed in meanwhile
+                                return;
+                            }
+
+                            provider.Refresh();
+                            ScheduleTimer(provider);
```

Comment:
Timer is initialized using AutoReset = false which leads to the timer Elapsed event only being thrown once and never again. Therefore a new timer needs to be scheduled to keep the "chain" of timers running until it is unregistered by the owning TimerBasedCredentialRefresher.

I did not alter this behavior from the version before this PR, see here:
https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/RabbitMQ.Client/client/api/ICredentialsRefresher.cs#L126

Reply:
Good to know the timer has to be disposed, next commit will include that.

From my perspective (I'm not a C# expert) using a chain of self triggering timers is a cleaner approach than a single persisting timer firing in a specific interval. This is because the elapse event method may take an indefinite amount of time to complete. So theoretically if the elapse method requires more time than the interval itself we get some type of congestion when multiple elapse methods run and update stuff. Of course in this scenario we would have other problems as well. Thus the elapse method may never take longer than 1/3 of the credentials lifetime (JWT for example).
Also see this article here for an more in-depth example (JS, but basically transferable to this use case here): https://reallifejs.com/brainchunks/repeated-events-timeout-or-interval/

Reply:
Yeah, that makes sense, it might also make sense to look into making the ICredentialsProvider properly async in 8.0 and then looking at better ways to do periodic refresh, i.e. PeriodicTimer or some maybe even a dumb refresh/delay loop might be enough.

Reply:
>  it might also make sense to look into making the ICredentialsProvider properly async in 8.0

No time like the present ... before version 7 ships!

Reply:
> No time like the present ... before version 7 ships!

Yeah, I  just took a good long look at it, there is quite a bit of change there, including figuring out the problem this PR solves, just with the added complexity of async and different target frameworks (only .NET 6 supports PeriodicTimer, which is the only real async capable timer, including the fact that we don't know the timer period until the token has been refreshed atleast once and it might change while the timer is running), then a lot of interfaces and related types need to be changed.
And probably a different concept for static credentials (basic auth) vs. dynamic stuff (oauth2).

I tried out the above here: 
https://github.com/Tornhoof/rabbitmq-dotnet-client/commit/8ec63c18126a98e08ab5d972d72f0a68ed113798

I personally don't know how much in use the OAuth2 stuff is, I've never used it with RabbitMQ, imho the OAuthClient requires quite a bit of overhaul (see my changes above), a lot of disposes are missing,  questionable use of System.Text.Json and the whole  sync over async code is questionable at best.

If you really want to release 7.0 this month, I'd recommend moving this to 8.

Reply:
I appreciate your input, and I'll discuss with @MarcialRosales and other Team RabbitMQ members next week.